### PR TITLE
2310 multi unit item] add api, service and repository

### LINKED ISSUE #2310: [Multi Unit Item] Add API, service and repository
Epic: #2078 

Some mock API was already added to query for unit variants (UnitVariantNode), further mock API would have been added in this issue: https://github.com/openmsupply/open-msupply/issues/2309.

This API needs to be implemented and requires, repository and service for item variants

### PR DESCRIPTION
**Pull request by roxy-dao**
Fixes #2310

# 👩🏻‍💻 What does this PR do? 
- Create `pack_unit` table.
- Logic to calculate `total_number_of_packs` or `total_number_of_lines` based on store's stock on hand to find the most used variant based on highest total.
- Graphql endpoint to get all pack units.

# 🧪 How has/should this change been tested? 
- [ ] Run tests? (Probably not the best test 😭 )

## 💌 Any notes for the reviewer?
Not sure if we need translations at this point since omSupply central hasn't been configured yet? Can create another issue for it ig

### CODE REVIEW

File: server/service/src/pack_unit.rs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        StockOnHandFilter::new().store_id(EqualFilter::equal_to(store_id)),
+    ))?;
+    let pack_units = PackUnitRowRepository::new(connection).load_all()?;
+
+    // Calculate the most used variant for each item and pack size by total number of packs
+    // if item has stock on hand else by empty line count. HashMap keys are (item_id, pack_size)
+    // for easier management of totals.
+    let mut total_number_of_packs: HashMap<(String, i32), f64> = HashMap::new();
+    let mut total_number_of_lines: HashMap<(String, i32), f64> = HashMap::new();
+    for stock_line in stock_lines {
```
Comment (roxy-dao) [inline server/service/src/pack_unit.rs:31]: Thought about doing all this in an SQL view but might be too complicated? 🤔 idk

File: client/packages/system/src/Item/api/hooks/useItemVariants/useUnitVariantList.ts [hunk 2]
```diff
@@ -4,7 +4,5 @@ import { useItemApi } from '../useItemApi';
 export const useUnitVariantList = () => {
   const api = useItemApi();
 
-  return useQuery(api.keys.unitVariantsList(), () =>
-    api.get.unitVariantsList()
-  );
+  return useQuery(api.keys.packUnits(), () => api.get.packUnits());
```
Comment (mark-prins) [inline client/packages/system/src/Item/api/hooks/useItemVariants/useUnitVariantList.ts:7]: No need to create a new method, the `api.get.packUnits` has no parameters, so you can just:
```suggestion
  return useQuery(api.keys.packUnits(), api.get.packUnits);
```

Comment (mark-prins) [inline client/packages/system/src/Item/api/hooks/useItemVariants/useUnitVariantList.ts:7]: Is there a reason that the folder is `useItemVariants` and the file is `useUnitVariantList`? 🤔

File: server/service/src/pack_unit.rs (hunk 1 above)
Comment (mark-prins) [reply server/service/src/pack_unit.rs:31]: we will need to test a large datafile, the login is already a slow process!
I'd prefer to have the logic in rust and keep the database clean - it's hard to maintain logic in two places in the codebase

**Review by mark-prins — DISMISSED** [review]
👍 LGTM!

File: client/packages/system/src/Item/api/operations.graphql [hunk 3]
```diff
@@ -237,23 +237,27 @@ query itemById($storeId: String!, $itemId: String!) {
...
   shortName
 }
 
-fragment UnitVariant on UnitVariantNode {
+fragment PackUnit on PackUnitNode {
   itemId
   mostUsedVariantId
-  variants {
-    ...Variant
+  packUnits {
```
Comment (mark-prins) [inline client/packages/system/src/Item/api/operations.graphql:250]: hasn't `mostUsedVariantId` been changed to `mostUsedPackSizeId`?

File: client/packages/system/src/Item/api/hooks/useItemVariants/useUnitVariantList.ts (hunk 2 above)
Comment (roxy-dao) [reply client/packages/system/src/Item/api/hooks/useItemVariants/useUnitVariantList.ts:7]: There is an issue for renaming... I jumped the gun and did it in places that I passed through 🙇🏻

File: server/service/src/pack_unit.rs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+                .and_modify(|e| *e += 1.0)
+                .or_insert(0.0);
+        }
+    }
+
+    // Find the most used variant id for each item based on total number of packs (if SOH) or total number of lines,
+    // and group all the pack units with their item
+    let mut pack_units_grouped: HashMap<String, (String, Vec<PackUnitRow>)> = HashMap::new();
+    for pack_unit in pack_units.clone() {
+        let item_id = pack_unit.item_id.clone();
```
Comment (clemens-msupply) [inline server/service/src/pack_unit.rs:55]:
```suggestion
    for pack_unit in &pack_units {
```

or `pack_units.iter()` works as well (previously it would do `pack_units.into_iter()`...)

File: server/service/src/pack_unit.rs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+        let most_used_pack_unit_id = pack_units
+            .iter()
+            .filter(|p| p.item_id == item_id)
+            .max_by(|a, b| {
```
Comment (clemens-msupply) [inline server/service/src/pack_unit.rs:60]: Found it very hard to follow, I restructured it a bit and wonder if this is better:

```rust
let ordering_by_packs = total_number_of_packs
                    .get(&(a.item_id.clone(), a.pack_size))
                    .unwrap_or(&0.0)
                    .partial_cmp(
                        total_number_of_packs
                            .get(&(b.item_id.clone(), b.pack_size))
                            .unwrap_or(&0.0),
                    )
                    .unwrap();
                let ordering_by_lines = total_number_of_lines
                    .get(&(a.item_id.clone(), a.pack_size))
                    .unwrap_or(&0.0)
                    .partial_cmp(
                        total_number_of_lines
                            .get(&(b.item_id.clone(), b.pack_size))
                            .unwrap_or(&0.0),
                    )
                    .unwrap();

                // order by number of pack and if equal order by number of lines:
                ordering_by_packs.then(ordering_by_lines)
```

File: server/service/src/pack_unit.rs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    // Find the most used variant id for each item based on total number of packs (if SOH) or total number of lines,
+    // and group all the pack units with their item
+    let mut pack_units_grouped: HashMap<String, (String, Vec<PackUnitRow>)> = HashMap::new();
+    for pack_unit in pack_units.clone() {
+        let item_id = pack_unit.item_id.clone();
+
+        let most_used_pack_unit_id = pack_units
+            .iter()
```
Comment (clemens-msupply) [inline server/service/src/pack_unit.rs:58]: That `most_used_pack_unit_id` isn't it?

File: server/service/src/pack_unit.rs [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+use repository::{
+    EqualFilter, PackUnitRow, PackUnitRowRepository, RepositoryError, StockLineFilter,
+    StockLineRepository, StockOnHandFilter, StockOnHandRepository, StorageConnection,
+};
+
+#[derive(Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct PackUnit {
+    pub item_id: String,
+    pub most_used_pack_unit_id: String,
+    pub pack_units: Vec<PackUnitRow>,
```
Comment (clemens-msupply) [inline server/service/src/pack_unit.rs:12]: Is this `most_used_pack_unit_id`? 

Think a doc comment for the struct fields is helpful here, especially for `pack_units` and `most_used_pack_size_id`...

The `PackUnit` struct name is a bit confusing because it contains a `pack_units` field. Maybe `ItemPackUnit`

File: server/service/src/pack_unit.rs (hunk 6 above)
Comment (roxy-dao) [reply server/service/src/pack_unit.rs:58]: Yup! Already been renamed (:

File: server/service/src/pack_unit.rs (hunk 7 above)
Comment (clemens-msupply) [reply server/service/src/pack_unit.rs:12]: Or PackUnitSummary or similar... can't think of a good name

File: server/service/src/pack_unit.rs [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+pub fn get_pack_units(
```
Comment (clemens-msupply) [inline server/service/src/pack_unit.rs:15]: Would the following doc comment be correct? Feel free to rephrase or move to PackUnit...

```rust
/// For a particular store, this method returns all pack units grouped by item_id and a reference to the most used pack unit for each item (see PackUnit return type).
```

File: server/service/src/pack_unit.rs [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+    connection: &StorageConnection,
+    store_id: &str,
+) -> Result<Vec<PackUnit>, RepositoryError> {
+    let stock_lines = StockLineRepository::new(connection)
+        .query_by_filter(StockLineFilter::new(), Some(store_id.to_string()))?;
+    let stock_on_hand = StockOnHandRepository::new(connection).query(Some(
+        StockOnHandFilter::new().store_id(EqualFilter::equal_to(store_id)),
+    ))?;
```
Comment (clemens-msupply) [inline server/service/src/pack_unit.rs:23]: Optimization: filter by available_stock_on_hand > 0 (if filter exist)

File: server/service/src/pack_unit.rs [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+    pub most_used_pack_unit_id: String,
+    pub pack_units: Vec<PackUnitRow>,
+}
+
+pub fn get_pack_units(
+    connection: &StorageConnection,
+    store_id: &str,
+) -> Result<Vec<PackUnit>, RepositoryError> {
+    let stock_lines = StockLineRepository::new(connection)
+        .query_by_filter(StockLineFilter::new(), Some(store_id.to_string()))?;
```
Comment (clemens-msupply) [inline server/service/src/pack_unit.rs:20]: This query is probably the most expensive one because it also returns joined rows which are not needed. So this could be optimized if needed, i.e. have a repo or repo method to only return the item_id + pack_size or a joined pack_unit + pack_size query...

Regarding your original question. You could create a view that gives you the data in a nicer format and do the rest of the logic here. The view columns could be like:

|item_id|pack_size|sum_available_stock_on_hand|total_number_of_stock_lines|

where the view is grouped by item_id and pack_size. This might be not too complicated. 

However, I am a bit with with Mark, it is probably simpler to understand when leaving it as it is, i.e. all logic in one place.

File: server/service/src/pack_unit.rs (hunk 5 above)
Comment (roxy-dao) [reply server/service/src/pack_unit.rs:60]: But we want to get the max value from those HashMaps 🤔

File: server/service/src/pack_unit.rs (hunk 8 above)
Comment (roxy-dao) [reply server/service/src/pack_unit.rs:15]: Correct~ have added

File: server/service/src/pack_unit.rs (hunk 9 above)
Comment (roxy-dao) [reply server/service/src/pack_unit.rs:23]: Filtering doesn't exist right now.. so might just leave this and if it does become a performance issue can create the filter

File: server/service/src/pack_unit.rs [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+#[derive(Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct ItemPackUnit {
+    pub item_id: String,
+    pub most_used_pack_unit_id: String,
+    pub pack_units: Vec<PackUnitRow>,
+}
+
+// For a particular store, this method returns all pack units grouped by item_id and a reference to the most
+// used pack unit for each item (see ItemPackUnit return type).
+pub fn get_pack_units(
```
Comment (clemens-msupply) [inline server/service/src/pack_unit.rs:17]: Please use `///`` slashes to make it a doc comment. For example, then the line shows when hovering where the function is used...

Comment (roxy-dao) [reply server/service/src/pack_unit.rs:17]: [185233d](https://github.com/openmsupply/open-msupply/pull/2357/commits/185233d72fa194dc1a26ebc6bf518b62ac5ec18b) XD

**Review by clemens-msupply — APPROVED** [review]
Changes look good thanks! Only looked at the service file though since Mark already reviewed the rest.
